Labels set on the command line don't override labels in Dockerfile

yongtang opened the pull request:
**\- What I did**

This fix tries to address the inconsistency in #22036 where labels set on the command line will not override labels specified in Dockerfile, but will override labels inherited from `FROM` images.

**\- How I did it**

The fix add a LABEL with command line options at the end of the processed Dockerfile so that command line options labels always override the LABEL in Dockerfiles (or through `FROM`).

**\- How to verify it**

An integration test has been added for test cases specified in #22036.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #22036.

NOTE: Some changes are from #22266 (@tiborvass).

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang commented:
@tiborvass @thaJeztah The pull request has been updated. Let me know if there are any issues.

tiborvass commented:
@yongtang i think the proper solution here is to not merge the label with the last config, but simply insert a LABEL instruction at the end. It will give less surprises.

yongtang commented:
Thanks @tiborvass. Let me update the pull request.

tiborvass commented:
@yongtang do you need help? What's your timezone? We might need to get the fix in fast if we want it for 1.11.1 :S let me know how you want it.

yongtang commented:
Thanks @tiborvass I was busy moments ago. I will try to see if I could get the fix in several hours. If I can not get it done quick I will let you know shortly.

tiborvass commented:
@yongtang any news?

yongtang commented:
Hi @tiborvass I am having some issue to locate a simple way to add additional lines of `LABEL foo=bar`. 

My previous understanding was that we could add `LABEL foo=bar` to the end of the Dockerfile or add the child to the end.
 
But the parsing of the Dockerfile is done in Parser and only `io.Reader` is passed. If we want to add additional lines then we need call either  `parseLine()` in build, or pass builder to `Parse()`. There will be quite some API changes.

I am wondering if I missed something, or there is a simple way to get it done?

mlaventure commented:
How about wrapping the file io.reader and when it returns EOF you'd just need to return a generated "LABEL" line then EOF.

tiborvass commented:
@yongtang we can insert it after parsing. let me show you.

tiborvass commented:
@yongtang we could export `builder/dockerfile/parser.parseLine` and call that in `builder/dockerfile/builder.go` before `for i, n := range b.dockerfile.Children`, and append the node to `b.dockerfile.Children` does that make sense?

yongtang commented:
Thanks @tiborvass let me give it a try.

tiborvass commented:
@yongtang 

```
line := "LABEL "
for k, v := range labels {
	line += fmt.Sprintf("%q=%q ", k, v)
}
_, node, err := parser.ParseLine(line)
```

using %q to quote user input, just in case :)

yongtang commented:
Thanks @tiborvass @mlaventure for the help. Just updated the pull request. I am not very familiar with the dockerfile's parser so it takes some time for me. Sorry for the delay and let me know if there are any issues.

yongtang commented:
There is a golint format issue in the previous commit. Just updated the current one.

yongtang commented:
@tiborvass @mlaventure Just updated the PR again to remove not used `addLabels()` altogether.

tiborvass commented:
@yongtang sorry i had to go away for a couple of hours and wasn't sure if you'd still be here so I opened https://github.com/docker/docker/pull/22266 against the 1.11 branch (not master) I really hope you don't mind. Your PR made me realize I forgot to remove `addLabels()` but then I realized you forgot to remove https://github.com/docker/docker/pull/22266/files#diff-f157a3a45b3e5d85aadff73bff1f5a7cL416

Also thanks for the test, there are a couple of nits i'll comment on, but I'll steal it :)

yongtang commented:
No problem @tiborvass and thanks for the help during the process.

tiborvass commented:
Feel free to update your PR :)

yongtang commented:
Thanks a lot @tiborvass.

tiborvass commented:
LGTM

thaJeztah commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -6894,3 +6894,62 @@ func (s *DockerRegistryAuthHtpasswdSuite) TestBuildWithExternalAuth(c *check.C)
+	res = inspectFieldJSON(c, name, "Config.Labels")
+	if res != expected {
+		c.Fatalf("Labels %s, expected %s", res, expected)
+	}
```
do we need test-cases with a command-line label *without* a value set? (`--label foo`, `--label foo=`)

tiborvass replied:
sure, but its not necessary for this pr to be merged

yongtang replied:
@tiborvass @thaJeztah In Docker 1.11.0 the command line options without value set will return an error:
```
$ sudo docker build -t --label=foo scratchy .
invalid value "--label=foo" for flag -t: Error parsing reference: "--label=foo" is not
a valid repository/tag
See 'docker build --help'.
```

```
$ sudo docker build -t --label=bar= scratchy .
invalid value "--label=bar=" for flag -t: Error parsing reference: "--label=bar=" is not
a valid repository/tag
See 'docker build --help'.
```

The error was handled in 
https://github.com/docker/docker/blob/v1.11.0/reference/reference.go#L51-L60
https://github.com/docker/docker/blob/v1.11.0/api/client/build.go#L284-L292

I will be out for about several hours but could add test cases for those error conditions after I am back.

yongtang replied:
Hi @tiborvass @thaJeztah my last comment was incorrect as I actually executed the incorrect command (had a schedule out so didn't check carefully, in hurry).

In Docker 1.11.0, here is the expected behavior:
(NOTE:  As is discussed in #22036, the command line label options will not be in use when the Dockerfile is only from scratch. The command line label options will override when Dockerfile is from another image. The following are the same but may looks like confusing.)
```
$ docker build -t scratchy --label foo= .

Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in f4c190a282ed
 ---> 4f8139a0fda2
Removing intermediate container f4c190a282ed
Successfully built 4f8139a0fda2
$ docker inspect --format "{{ json .Config.Labels }}" scratchy

{"bar":"from-dockerfile","foo":""}
(We passed a --label foo=, which is treated as foo="")
```

```
$ docker build -t scratchy --label bar .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in 81979aaafc79
 ---> 26ddc83918ea
Removing intermediate container 81979aaafc79
Successfully built 26ddc83918ea
$ docker inspect --format "{{ json .Config.Labels }}" scratchy
{"bar":"from-dockerfile"}

We passed --label bar which is treated as --label bar="", however, 
docker will not override here so no effect.
```

```
$ docker build -t scratchy --label bar .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in 32c8cce82af3
 ---> 4d5aaff2569b
Removing intermediate container 32c8cce82af3
Successfully built 4d5aaff2569b

$ docker build -t scratchy2 --label foo .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratchy
 ---> 4d5aaff2569b
Step 2 : LABEL bar from-dockerfile2
 ---> Running in 6f4af1e6fdcb
 ---> a17708718028
Removing intermediate container 6f4af1e6fdcb
Successfully built a17708718028
$ docker inspect --format "{{ json .Config.Labels }}" scratchy2
{"bar":"from-dockerfile2","foo":""}
We passed --label foo which is equal to --label foo="", and in this case,
there is an override so we see "foo": ""
```

yongtang replied:
Hi @tiborvass @thaJeztah Based on the above, my understanding is that, a command line option of `--label foo=` or `--label bar` means `--label foo=""` and `--label bar=""`, and they will ALWAY override `LABEL` in Dockerfile.

I will add a couple of test cases to cover these scenarios. But do let me know if we have other thought on the behavior of the command line label options.

yongtang replied:
@tiborvass @thaJeztah I just updated the pull request with added tests for `--label foo` and `--label bar`. Please let me know if there are any issues.

thaJeztah replied:
> Based on the above, my understanding is that, a command line option of --label foo= or --label bar means --label foo="" and --label bar="", and they will ALWAY override LABEL in Dockerfile.

Yes, I think that's the right expectation

yongtang committed:
Labels set on the command line always override labels in Dockerfile

This fix tries to address the inconsistency in #22036 where labels
set on the command line will not override labels specified in
Dockerfile, but will override labels inherited from `FROM` images.

The fix add a LABEL with command line options at the end of the
processed Dockerfile so that command line options labels always
override the LABEL in Dockerfiles (or through `FROM`).

An integration test has been added for test cases specified in #22036.

This fix fixes #22036.

NOTE: Some changes are from #22266 (@tiborvass).

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

vdemeester commented:
LGTM 🐰
